From a0bc4a30eaf4e41c2d5237867c5f9e93a7d1d363 Mon Sep 17 00:00:00 2001
From: allen yan <yanwei@marvell.com>
Date: Fri, 30 Dec 2016 22:09:18 +0800
Subject: [PATCH 1283/2241] uart: a3700: add multiple uart port support

- The uart driver can only support single uart port now.
  This patch updated the existing driver to extend the
  maximum port number up to 2. The driver is supposed to
  take the uart alias in device-tree as the port index.

Change-Id: If4d114a38d5bf6973494665dcec7dbfd6c41fc53
Signed-off-by: allen yan <yanwei@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/35173
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Wilson Ding <dingwei@marvell.com>
---
 drivers/tty/serial/mvebu-uart.c | 15 ++++++++++++---
 1 file changed, 12 insertions(+), 3 deletions(-)

diff --git a/drivers/tty/serial/mvebu-uart.c b/drivers/tty/serial/mvebu-uart.c
index 28a29d1..c97556c 100644
--- a/drivers/tty/serial/mvebu-uart.c
+++ b/drivers/tty/serial/mvebu-uart.c
@@ -86,7 +86,7 @@
 
 #define UART_BRDV		0x10
 
-#define MVEBU_NR_UARTS		1
+#define MVEBU_NR_UARTS		2
 
 #define MVEBU_UART_TYPE		"mvebu-uart"
 #define DRIVER_NAME		"mvebu-serial"
@@ -557,7 +557,15 @@ static int mvebu_uart_probe(struct platform_device *pdev)
 		return -EINVAL;
 	}
 
-	port = &mvebu_uart_ports[0];
+	if (pdev->dev.of_node)
+		pdev->id = of_alias_get_id(pdev->dev.of_node, "serial");
+
+	if (pdev->id >= MVEBU_NR_UARTS) {
+		dev_err(&pdev->dev, "exceed max uart ports\n");
+		return -EINVAL;
+	}
+
+	port = &mvebu_uart_ports[pdev->id];
 
 	spin_lock_init(&port->lock);
 
@@ -569,7 +577,7 @@ static int mvebu_uart_probe(struct platform_device *pdev)
 	port->fifosize   = 32;
 	port->iotype     = UPIO_MEM32;
 	port->flags      = UPF_FIXED_PORT;
-	port->line       = 0; /* single port: force line number to  0 */
+	port->line       = pdev->id;
 
 	port->irq        = irq->start;
 	port->irqflags   = IRQF_SHARED;
@@ -592,6 +600,7 @@ static int mvebu_uart_probe(struct platform_device *pdev)
 	ret = uart_add_one_port(&mvebu_uart_driver, port);
 	if (ret)
 		return ret;
+
 	return 0;
 }
 
-- 
2.7.4

